ntoskrnl/mm/ARM3/wslist.cpp

```diff
@@ -0,0 +1,484 @@
+
+    PMMPTE PointerPte = MiAddressToPte(Address);
+
+    /* Make sure we are adding a paged-in address */
+    ASSERT(PointerPte->u.Hard.Valid == 1);
+    PMMPFN Pfn1 = MiGetPfnEntry(PFN_FROM_PTE(PointerPte));
+    ASSERT(Pfn1->u3.e1.PageLocation == ActiveAndValid);
+
+    /* Shared pages not supported yet */
+    ASSERT(Pfn1->u1.WsIndex == 0);
+    ASSERT(Pfn1->u3.e1.PrototypePte == 0);
+
+    /* Nor are "ROS PFN" */
+    ASSERT(MI_IS_ROS_PFN(Pfn1) == FALSE);
+
+    Pfn1->u1.WsIndex = GetFreeWsleIndex(WsList);
+    MMWSLENTRY& NewWsle = WsList->Wsle[Pfn1->u1.WsIndex].u1.e1;
+    NewWsle.VirtualPageNumber = reinterpret_cast<ULONG_PTR>(Address) >> PAGE_SHIFT;
+    NewWsle.Protection = Protection;
+    NewWsle.Direct = 1;
+    NewWsle.Hashed = 0;
+    NewWsle.LockedInMemory = 0;
+    NewWsle.LockedInWs = 0;
+    NewWsle.Age = 0;
+    NewWsle.Valid = 1;
+
+    Vm->WorkingSetSize += PAGE_SIZE;
+    if (Vm->WorkingSetSize > Vm->PeakWorkingSetSize)
+        Vm->PeakWorkingSetSize = Vm->WorkingSetSize;
+}
+
+VOID
+NTAPI
+MiRemoveFromWorkingSetList(
+    _Inout_ PMMSUPPORT Vm,
+    _In_ PVOID Address)
+{
+    RemoveFromWsList(Vm->VmWorkingSetList, Address);
+
+    Vm->WorkingSetSize -= PAGE_SIZE;
+}
+
+_Requires_exclusive_lock_held_(WorkingSet->WorkingSetMutex)
+VOID
+NTAPI
+MiInitializeWorkingSetList(_Inout_ PMMSUPPORT WorkingSet)
+{
+    PMMWSL WsList = WorkingSet->VmWorkingSetList;
+
+    /* Initialize some fields */
+    WsList->FirstFree = ULONG_MAX;
+    WsList->Wsle = reinterpret_cast<PMMWSLE>(WsList + 1);
+    WsList->LastEntry = 0;
+    /* The first page is already allocated */
+    WsList->LastInitializedWsle = (PAGE_SIZE - sizeof(*WsList)) / sizeof(MMWSLE);
+
+    /* Insert the address we already know: our PDE base and the Working Set List */
+    if (MI_IS_PROCESS_WORKING_SET(WorkingSet))
+    {
+        ASSERT(WorkingSet->VmWorkingSetList == MmWorkingSetList);
+#if _MI_PAGING_LEVELS == 4
+        MiInsertInWorkingSetList(WorkingSet, (PVOID)PXE_BASE, 0U);
+#elif _MI_PAGING_LEVELS == 3
+        MiInsertInWorkingSetList(WorkingSet, (PVOID)PPE_BASE, 0U);
+#elif _MI_PAGING_LEVELS == 2
+        MiInsertInWorkingSetList(WorkingSet, (PVOID)PDE_BASE, 0U);
+#endif
+    }
+
+#if _MI_PAGING_LEVELS == 4
+    MiInsertInWorkingSetList(WorkingSet, MiAddressToPpe(WorkingSet->VmWorkingSetList), 0UL);
+#endif
+#if _MI_PAGING_LEVELS >= 3
+    MiInsertInWorkingSetList(WorkingSet, MiAddressToPde(WorkingSet->VmWorkingSetList), 0UL);
+#endif
+    MiInsertInWorkingSetList(WorkingSet, (PVOID)MiAddressToPte(WorkingSet->VmWorkingSetList), 0UL);
+    MiInsertInWorkingSetList(WorkingSet, (PVOID)WorkingSet->VmWorkingSetList, 0UL);
+
+    /* From now on, every added page can be trimmed at any time */
+    WsList->FirstDynamic = WsList->LastEntry;
+
+    /* We can add this to our list */
+    ExInterlockedInsertTailList(&MmWorkingSetExpansionHead, &WorkingSet->WorkingSetExpansionLinks, &MmExpansionLock);
+}
+
+VOID
+NTAPI
+MmWorkingSetManager(VOID)
```

Comment:
You don't want to enable it in this PR?

Reply:
There's no point, no body adds anything to WS lists yet :-)